File: BowlingGame/ruby/unrefactored/game.rb
# frozen_string_literal: true

class Game
  attr_reader :record

  def initialize
    @record = []
  end

  def roll(pins)
    @record << pins
  end

  def score
    total = 0
    roll = 0
    10.times do
      if @record[roll] == 10
        total += 10 + @record[roll+1] + @record[roll+2]
        roll += 1
      else
        total += @record[roll] + @record[roll+1]
        if @record[roll] + @record[roll+1] == 10
          total += @record[roll+2]
        end
        roll += 2
      end
    end
    total
  end
end

File: BowlingGame/csharp-xunit/csharp-xunit/Game.cs
using System.Collections.Generic;

namespace csharp_xunit
{
    public class Game
    {
        private List<int> results = new List<int>();

        public int Score()
        {
            int total = 0;
            int nextItem = 0;
            for (int frame = 0; frame < 10; frame++)
            {
                total += Get(nextItem);
                bool isStrike = Get(nextItem) == 10;
                if (isStrike)
                {
                    total += Get(nextItem + 1) + Get(nextItem + 2);
                    nextItem += 1;
                }
                else
                {
                    total += Get(nextItem + 1);
                    bool isSpare = Get(nextItem) + Get(nextItem + 1) == 10;
                    if (isSpare)
                    {
                        total += Get(nextItem + 2);
                    }
                    nextItem += 2;
                }
            }
            return total;
        }

        public void Roll(int roll) => results.Add(roll);

        private int Get(int i)
        {
            // with Linq, could do results.ElementAtOrDefault(i);
            return (i < results.Count) ? results[i] : 0;
        }
    }
}
